cmd/compose/attach.go

```diff
@@ -0,0 +1,82 @@
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package compose
+
+import (
+	"context"
+
+	"github.com/docker/cli/cli/command"
+	"github.com/docker/cli/cli/command/container"
+	"github.com/docker/compose/v2/pkg/api"
+	"github.com/spf13/cobra"
+)
+
+func attachCommand(p *ProjectOptions, dockerCli command.Cli, backend api.Service) *cobra.Command {
+	// uses NewAttachCommand from `docker container attach`:
+	// - lookup container by service name (and optionally index)
+	// - invoke underlying `container attach` with container ID
+
+	// FYI changes to the underlying NewAttachCommand could require changes here, i.e. help messages or options added that don't make sense with `compose attach`
+	// - That said, the command flags/behavior are pretty stable
+
+	cmd := container.NewAttachCommand(dockerCli)
+	containerAttachRunE := cmd.RunE
+
+	// Options:
+	// inherits `container attach` options: --detach-keys, --no-stdin, --sig-proxy, --dry-run
+	// additional options specific to compose (in this case to lookup a container):
+	opts := api.LookupContainerOptions{}
+	cmd.Flags().IntVar(&opts.Index, "index", 1, "index of the container if service has multiple replicas")
+
+	// Help messages: (override to name first arg SERVICE instead of CONTAINER):
+	cmd.Use = "attach [OPTIONS] SERVICE"
+	cmd.Short = "Attach local standard input, output, and error streams to a service's running container"
+
+	// clear alias for `docker attach`
+	cmd.Annotations = map[string]string{}
+
+	// todo is this for completions or validation? or both?
+	cmd.ValidArgsFunction = completeServiceNames(dockerCli, p)
+
+	// Args: why require a service name?
+	// - It's possible to make `compose attach` work without specifying a service (i.e. just use first container).
+	// - But, I don't see that convention used in other similar compose commands, i.e. `compose exec` which targets a single container too and requires the service name.
+	// - Also, the underlying `container attach` command requires 1 arg so it would have to be patched to ignore no args
+	// - And, other compose commands optionally take a service name, i.e. `compose logs`, but the behavior there is to view logs of all containers not just one.
+
+	cmd.RunE = Adapt(func(ctx context.Context, args []string) error {
+		projectName, err := p.toProjectName(dockerCli)
+		if err != nil {
+			return err
+		}
+
+		if len(args) > 0 {
+			opts.Service = args[0]
+		}
+
+		containerID, err := backend.LookupContainer(ctx, projectName, opts)
```

Comment:
please mimic the `exec` commands, and have api.Backend to declare a new `Attach` command, then rely on docker/cli  for implementation if relevant.
This for sure requires to duplicate flags definition, but those don't change much and this is a general UX-homogeneity issue we should address in a global way

Reply:
Ok so you would prefer that I not reuse `NewAttachCommand` and instead hook into `RunAttach` like `exec` does with `RunExec`?

Reply:
And btw, if I hook into `RunAttach` I can easily add back the ability to `docker compose attach` (w/o service name) and have it grab the first container in the project. I removed that when using `NewAttachCommand` due to some friction with overriding validation.

Reply:
Yes indeed, I'd prefer not to rely on `NewAttachCommand` here (sharing command and flags definition with docker CLI is a larger effort we need to investigate)

> I can easily add back the ability to docker compose attach (w/o service name) and have it grab the first container in the project

please don't, containers in a project are not ordered, so this won't be deterministic. Better rely on the same selection mechanism used by `compose exec` to enforce an homogeneous UX